### PR TITLE
ETL Orchestrator

### PR DESCRIPTION
**Pull request by christianhelle**
This introduces modules containing reusable components for implementing ETL operations in Python libraries. I included documentation briefly describing the Orchestrated ETL design pattern together with generous code examples

The star of the show in this PR is the `Orchestration` fluent interface:

```python
from atc.etl import Extractor, Transformer, Loader, Orchestration

(Orchestration
    .extract_from(Extractor())
    .transform_with(Transformer())
    .load_into(Loader())
    .build()
    .execute())
```

## Usage examples:

Here are some example usages and implementations of the ETL class provided

### Example-1

```python
from pyspark.sql import DataFrame
from pyspark.sql.types import StructType, StructField, IntegerType, StringType
import pyspark.sql.functions as f

from atc.etl import Extractor, Transformer, Loader, Orchestration
from atc.spark import Spark


class GuitarExtractor(Extractor):
    def read(self) -> DataFrame:
        return Spark.get().createDataFrame(
            Spark.get().sparkContext.parallelize([
                ('1', 'Fender', 'Telecaster', '1950'),
                ('2', 'Gibson', 'Les Paul', '1959'),
                ('3', 'Ibanez', 'RG', '1987')
            ]),
            StructType([
                StructField('id', StringType()),
                StructField('brand', StringType()),
                StructField('model', StringType()),
                StructField('year', StringType()),
            ]))


class BasicTransformer(Transformer):
    def process(self, df: DataFrame) -> DataFrame:
        print('Current DataFrame schema')
        df.printSchema()

        df = df.withColumn('id', f.col('id').cast(IntegerType()))
        df = df.withColumn('year', f.col('year').cast(IntegerType()))

        print('New DataFrame schema')
        df.printSchema()
        return df


class NoopLoader(Loader):
    def save(self, df: DataFrame) -> DataFrame:
        df.write.format('noop').mode('overwrite').save()
        return df


print('ETL Orchestrator using a single simple transformer')
etl = (Orchestration
       .extract_from(GuitarExtractor())
       .transform_with(BasicTransformer())
       .load_into(NoopLoader())
       .build())
result = etl.execute()
result.printSchema()
result.show()
```

The code above produces the following output:

```
Original DataFrame schema
root
 |-- id: string (nullable = true)
 |-- brand: string (nullable = true)
 |-- model: string (nullable = true)
 |-- year: string (nullable = true)

New DataFrame schema
root
 |-- id: integer (nullable = true)
 |-- brand: string (nullable = true)
 |-- model: string (nullable = true)
 |-- year: integer (nullable = true)

+---+------+----------+----+
| id| brand|     model|year|
+---+------+----------+----+
|  1|Fender|Telecaster|1950|
|  2|Gibson|  Les Paul|1959|
|  3|Ibanez|        RG|1987|
+---+------+----------+----+
```

### Example-2

Using the [code above](#Example-1) as reference, the transformation code can be improved to be more generic and reused for other operations. 
Here's an example of implementing a `Transformer` that is reused to change the data type of a given column,
where the column name is parameterized

```python
import pyspark.sql.functions as f
from pyspark.sql import DataFrame
from pyspark.sql.types import StructType, StructField, IntegerType, StringType

from atc.etl import Extractor, Transformer, Loader, Orchestration
from atc.spark import Spark


class GuitarExtractor(Extractor):
    def read(self) -> DataFrame:
        return Spark.get().createDataFrame(
            Spark.get().sparkContext.parallelize([
                ('1', 'Fender', 'Telecaster', '1950'),
                ('2', 'Gibson', 'Les Paul', '1959'),
                ('3', 'Ibanez', 'RG', '1987')
            ]),
            StructType([
                StructField('id', StringType()),
                StructField('brand', StringType()),
                StructField('model', StringType()),
                StructField('year', StringType()),
            ]))


class IntegerColumnTransformer(Transformer):
    def __init__(self, col_name: str):
        self.col_name = col_name

    def process(self, df: DataFrame) -> DataFrame:
        df = df.withColumn(self.col_name, f.col(self.col_name).cast(IntegerType()))
        return df


class NoopLoader(Loader):
    def save(self, df: DataFrame) -> DataFrame:
        df.write.format('noop').mode('overwrite').save()
        return df


print('ETL Orchestrator using multiple transformers')
etl = (Orchestration
       .extract_from(GuitarExtractor())
       .transform_with(IntegerColumnTransformer('id'))
       .transform_with(IntegerColumnTransformer('year'))
       .load_into(NoopLoader())
       .build())
result = etl.execute()
result.printSchema()
result.show()
```

### Example-3

There are scenarios that you might have to ingest data from multiple data sources and merge them into a 
single dataframe. Here's an example of have multiple `Extractor` implementation encapsulated in an instance of 
`DelegatingExtractor` and applying transformations using the `MultiInputTransformer`

The `read()` function in `DelegatingExtractor` will return a dictionary that uses the type name of the `Extractor` 
as the key, and a `DataFrame` as its value

`MultiInputTransformer` provides the function `process_many(dataset: {})` and returns a single `DataFrame`

```python
import pyspark.sql.functions as f
from pyspark.sql import DataFrame
from pyspark.sql.types import StructType, StructField, StringType

from atc.etl import Extractor, Loader, MultiInputTransformer, Orchestration
from atc.spark import Spark


class AmericanGuitarExtractor(Extractor):
    def read(self) -> DataFrame:
        return Spark.get().createDataFrame(
            Spark.get().sparkContext.parallelize([
                ('1', 'Fender', 'Telecaster', '1950'),
                ('2', 'Gibson', 'Les Paul', '1959')
            ]),
            StructType([
                StructField('id', StringType()),
                StructField('brand', StringType()),
                StructField('model', StringType()),
                StructField('year', StringType()),
            ]))


class JapaneseGuitarExtractor(Extractor):
    def read(self) -> DataFrame:
        return Spark.get().createDataFrame(
            Spark.get().sparkContext.parallelize([
                ('3', 'Ibanez', 'RG', '1987'),
                ('4', 'Takamine', 'Pro Series', '1959')
            ]),
            StructType([
                StructField('id', StringType()),
                StructField('brand', StringType()),
                StructField('model', StringType()),
                StructField('year', StringType()),
            ]))


class CountryOfOriginTransformer(MultiInputTransformer):
    def process_many(self, dataset: {}) -> DataFrame:
        usa_df = dataset['AmericanGuitarExtractor'].withColumn('country', f.lit('USA'))
        jap_df = dataset['JapaneseGuitarExtractor'].withColumn('country', f.lit('Japan'))
        return usa_df.union(jap_df)


class NoopLoader(Loader):
    def save(self, df: DataFrame) -> DataFrame:
        df.write.format('noop').mode('overwrite').save()
        return df


print('ETL Orchestrator using multiple extractors')
etl = (Orchestration
       .extract_from(AmericanGuitarExtractor())
       .extract_from(JapaneseGuitarExtractor())
       .transform_with(CountryOfOriginTransformer())
       .load_into(NoopLoader())
       .build())
result = etl.execute()
result.printSchema()
result.show()
```

The code above produces the following output:

```
root
 |-- id: string (nullable = true)
 |-- brand: string (nullable = true)
 |-- model: string (nullable = true)
 |-- year: string (nullable = true)
 |-- country: string (nullable = false)

+---+--------+----------+----+-------+
| id|   brand|     model|year|country|
+---+--------+----------+----+-------+
|  1|  Fender|Telecaster|1950|    USA|
|  2|  Gibson|  Les Paul|1959|    USA|
|  3|  Ibanez|        RG|1987|  Japan|
|  4|Takamine|Pro Series|1959|  Japan|
+---+--------+----------+----+-------+
```

### Example-4

Here's an example of data raw ingestion without applying any transformations

```python
from pyspark.sql import DataFrame
from pyspark.sql.types import StructType, StructField, StringType

from atc.etl import Extractor, Loader, Orchestration
from atc.spark import Spark


class GuitarExtractor(Extractor):
    def read(self) -> DataFrame:
        return Spark.get().createDataFrame(
            Spark.get().sparkContext.parallelize([
                ('1', 'Fender', 'Telecaster', '1950'),
                ('2', 'Gibson', 'Les Paul', '1959'),
                ('3', 'Ibanez', 'RG', '1987')
            ]),
            StructType([
                StructField('id', StringType()),
                StructField('brand', StringType()),
                StructField('model', StringType()),
                StructField('year', StringType()),
            ]))


class NoopLoader(Loader):
    def save(self, df: DataFrame) -> DataFrame:
        df.write.format('noop').mode('overwrite').save()
        return df


print('ETL Orchestrator with no transformations')
etl = (Orchestration
       .extract_from(GuitarExtractor())
       .load_into(NoopLoader())
       .build())
result = etl.execute()
result.printSchema()
result.show()
```

### Example-5

Here's an example of writing the transformed data into multiple destinations

```python
import pyspark.sql.functions as f
from pyspark.sql import DataFrame
from pyspark.sql.types import StructType, StructField, IntegerType, StringType

from atc.etl import Extractor, Transformer, Loader, Orchestration
from atc.spark import Spark


class GuitarExtractor(Extractor):
    def read(self) -> DataFrame:
        return Spark.get().createDataFrame(
            Spark.get().sparkContext.parallelize([
                ('1', 'Fender', 'Telecaster', '1950'),
                ('2', 'Gibson', 'Les Paul', '1959'),
                ('3', 'Ibanez', 'RG', '1987')
            ]),
            StructType([
                StructField('id', StringType()),
                StructField('brand', StringType()),
                StructField('model', StringType()),
                StructField('year', StringType()),
            ]))


class BasicTransformer(Transformer):
    def process(self, df: DataFrame) -> DataFrame:
        print('Current DataFrame schema')
        df.printSchema()

        df = df.withColumn('id', f.col('id').cast(IntegerType()))
        df = df.withColumn('year', f.col('year').cast(IntegerType()))

        print('New DataFrame schema')
        df.printSchema()
        return df


class NoopSilverLoader(Loader):
    def save(self, df: DataFrame) -> DataFrame:
        df.write.format('noop').mode('overwrite').save()
        return df


class NoopGoldLoader(Loader):
    def save(self, df: DataFrame) -> DataFrame:
        df.write.format('noop').mode('overwrite').save()
        return df


print('ETL Orchestrator using multiple loaders')
etl = (Orchestration
       .extract_from(GuitarExtractor())
       .transform_with(BasicTransformer())
       .load_into(NoopSilverLoader())
       .load_into(NoopGoldLoader())
       .build())
result = etl.execute()
result.printSchema()
result.show()
```

### CODE REVIEW

File: src/atc/etl/extractor.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+class DelegatingExtractor(Extractor):
+    def __init__(self, inner_extractors: [Extractor]):
+        super().__init__()
+        self.inner_extractors = inner_extractors
+
+    def get_extractors(self) -> [Extractor]:
+        return self.inner_extractors
+
+    def read(self) -> dict:
```
Comment (JeppeBlixen) [inline src/atc/etl/extractor.py:23]: Mabey adding dict typing:
from typing import Dict
Dict[str, DataFrame]

File: src/atc/etl/orchestrator.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    @staticmethod
+    def create_for_multiple_transformers(extractor: Extractor,
+                                         transformers: [Transformer],
+                                         loader: Loader) -> MultipleTransformOrchestrator:
+        return MultipleTransformOrchestrator(extractor, DelegatingTransformer(transformers), loader)
+
+    @staticmethod
+    def create_for_raw_ingestion(extractor: Extractor,
+                                 loader: Loader) -> NoTransformOrchestrator:
+        return NoTransformOrchestrator(extractor, loader)
```
Comment (mrmasterplan) [inline src/atc/etl/orchestrator.py:74]: Why are these not the `__init__` functions of the respective classes. That would seem much more standard python to do.

File: src/atc/etl/transformer.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class MultiInputTransformer:
+    @abstractmethod
+    def process_many(self, dataset: {}) -> DataFrame:
+        pass
+
+
+class DelegatingTransformer(Transformer):
+    def __init__(self, inner_transformers: [Transformer]):
```
Comment (mrmasterplan) [inline src/atc/etl/transformer.py:19]: I didn't know `[Transformer]` was allowable as a type hint. Is this equivalent to `typing.List[Transformer]`?

**Review by mrmasterplan — COMMENTED** [review]
Overall awesome work!! I just don't think you should use a factory pattern. It is quite unnecessary here.
My proposal: Either make these `__init__` methods of the respective classes, or make it a single create method that inspects its members to deduce the type of orchestrator. If you force me to call the right method of the factory, you might as well make them into `__init__` methods. That's just my my opinion, and I'm prepared to be overruled if Chris and Jeppe agree with the current implementation.

File: src/atc/etl/orchestrator.py (hunk 2 above)
Comment (JeppeBlixen) [reply src/atc/etl/orchestrator.py:74]: Maybe these function could be merged to a single create, with some implementation logic?
def create(extractor: typing.Union[Extractor, typing.List[Extractor]],
       transformer: typing.Optional[typing.Union[Extractor, typing.List[Transformer]]],
       loader: Loader) -> Orchestrator:

File: src/atc/etl/transformer.py (hunk 3 above)
Comment (JeppeBlixen) [reply src/atc/etl/transformer.py:19]: I am also used to use typing.List[Transformer]

**Comment by JeppeBlixen** [comment]
Super great work! I basically have the same comment as Simon.

File: src/atc/etl/extractor.py (hunk 1 above)
Comment (christianhelle) [reply src/atc/etl/extractor.py:23]: Good idea! thanks

**Comment by christianhelle** [comment]
> Overall awesome work!! I just don't think you should use a factory pattern. It is quite unnecessary here. My proposal: Either make these `__init__` methods of the respective classes, or make it a single create method that inspects its members to deduce the type of orchestrator. If you force me to call the right method of the factory, you might as well make them into `__init__` methods. That's just my my opinion, and I'm prepared to be overruled if Chris and Jeppe agree with the current implementation.

I actually agree that the `OrchestratorFactory` is not a proper usage of the abstract factory pattern. 

What I really want is a to be able to construct ETL Orchestrators without knowing the specifics of which Orchestrator implementation to use

Maybe a fluent interface like this:

**Example-1** - Basic transformation using single source

```python
(Ochestration
    .extract_using(GuitarExtractor())
    .transform_with(BasicTransformer())
    .load_into(NoopLoader())
    .execute())
```

**Example-2** - Multiple transformations of single source

```python
(Ochestration
    .extract_using(GuitarExtractor())
    .transform_with(IntegerColumnTransformer('id'))
    .transform_with(IntegerColumnTransformer('year'))
    .load_into(NoopLoader())
    .execute())
```

**Example-3** - Single transformation on multiple sources

```python
(Ochestration
    .extract_using(AmericanGuitarExtractor())
    .extract_using(JapaneseGuitarExtractor())
    .transform_with(CountryOfOriginTransformer)
    .load_into(NoopLoader())
    .execute())
```

**Example-4** - Raw data ingestion with no transformation

```python
(Ochestration
    .extract_using(GuitarExtractor())
    .load_into(NoopLoader())
    .execute())
```

But I honestly didn't know how to write it in python at the time :D

**Comment by christianhelle** [comment]
> Either make these `__init__` methods of the respective classes, or make it a single create method that inspects its members to deduce the type of orchestrator. If you force me to call the right method of the factory, you might as well make them into `__init__` methods. That's just my my opinion, and I'm prepared to be overruled if Chris and Jeppe agree with the current implementation.


Having a single `create` function would be the correct implementation of the abstract factory pattern but it's not very clear for the caller to know how the `Orchestrator` will be created

I really like the fact that these functions

```python
def create(extractor: Extractor, transformer: Transformer, loader: Loader):
def create_for_raw_ingestion(extractor: Extractor, loader: Loader):
def create_for_multiple_sources(extractors: [Extractor], transformer: MultiInputTransformer, loader: Loader):
def create_for_multiple_transformers(extractor: Extractor, transformers: [Transformer], loader: Loader):
```

are named exactly after what they do

File: src/atc/etl/transformer.py (hunk 3 above)
Comment (christianhelle) [reply src/atc/etl/transformer.py:19]: > I didn't know `[Transformer]` was allowable as a type hint. Is this equivalent to `typing.List[Transformer]`?

I have honestly no clue. I'm still trying to learn, but there were no errors or warnings in PyCharm. It seems that the internet favors `List[T]` and if its on the internet it must be correct :D

File: src/atc/etl/orchestrator.py (hunk 2 above)
Comment (christianhelle) [reply src/atc/etl/orchestrator.py:74]: > Why are these not the `__init__` functions of the respective classes. That would seem much more standard python to do.

The `__init__` functions of the respective classes have the same arguments

Comment (christianhelle) [reply src/atc/etl/orchestrator.py:74]: > There could also be multiple loaders, in cases where data is save to multiple places.

Ah yes! I forgot about that scenario. I would like to implement that feature as well

File: src/atc/etl/transformer.py (hunk 3 above)
Comment (christianhelle) [reply src/atc/etl/transformer.py]: Done

**Comment by christianhelle** [comment]
Thanks for the builder pattern implementation @mrmasterplan 

I'll update the docs and example code

**Comment by christianhelle** [comment]
> > Either make these `__init__` methods of the respective classes, or make it a single create method that inspects its members to deduce the type of orchestrator. If you force me to call the right method of the factory, you might as well make them into `__init__` methods. That's just my my opinion, and I'm prepared to be overruled if Chris and Jeppe agree with the current implementation.
> 
> Having a single `create` function would be the correct implementation of the abstract factory pattern but it's not very clear for the caller to know how the `Orchestrator` will be created
> 
> I really like the fact that these functions
> 
> ```python
> def create(extractor: Extractor, transformer: Transformer, loader: Loader):
> def create_for_raw_ingestion(extractor: Extractor, loader: Loader):
> def create_for_multiple_sources(extractors: [Extractor], transformer: MultiInputTransformer, loader: Loader):
> def create_for_multiple_transformers(extractor: Extractor, transformers: [Transformer], loader: Loader):
> ```
> 
> are named exactly after what they do

The new `Orchestration` fluent interface replaces the `OrchestratorFactory`

**Review by mrmasterplan — APPROVED** [review]

**Review by JeppeBlixen — APPROVED** [review]

**Pull request merged**